File: dart/leet_code/1249.dart
/// Dart implementation of the LeetCode problem 1249: Minimum Remove to Make Valid Parentheses.
///
/// This module contains a class `Solution` with a method `minRemoveToMakeValid`.
/// The `minRemoveToMakeValid` method takes in one parameter:
/// - `s`: A string representing the input string of parentheses.
///
/// The method returns a string representing the valid parentheses string after removing the minimum number of parentheses.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `s` is a string of parentheses.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.

class Solution {
  /// Removes the minimum number of parentheses to make the input string valid.
  ///
  /// This method uses a stack to keep track of the indices of the open parentheses.
  /// It iterates over each character in the input string. If the character is an open parenthesis, it adds its index to the stack.
  /// If the character is a close parenthesis and the stack is empty, it adds its index to the remove list.
  /// If the character is a close parenthesis and the stack is not empty, it removes the last index from the stack.
  /// After the iteration, it adds the indices in the stack to the remove list and sorts the list.
  /// Then, it constructs the result string by skipping the characters at the indices in the remove list.
  ///
  /// @param s The input string of parentheses.
  /// @return A string representing the valid parentheses string after removing the minimum number of parentheses.
  String minRemoveToMakeValid(String s) {
    List<int> stack = [];
    List<int> remove = [];
    for (int i = 0; i < s.length; i++) {
      if (s[i] == '(') {
        stack.add(i);
      } else if (s[i] == ')') {
        if (stack.isEmpty) {
          remove.add(i);
        } else {
          stack.removeLast();
        }
      }
    }
    stack.addAll(remove);
    stack.sort();
    String result = '';
    for (int i = 0; i < s.length; i++) {
      if (stack.isNotEmpty && stack.first == i) {
        stack.removeAt(0);
      } else {
        result += s[i];
      }
    }
    return result;
  }
}

File: javascript/leet_code/1249.mjs
/**
 * @file This module removes invalid parentheses from a string to make it valid.
 * @module leet_code/1249
 */

/**
 * Removes the minimum number of invalid parentheses in order to make the input string valid.
 *
 * The function uses a stack to keep track of the indices of the open parentheses. It also keeps
 * track of the indices of the unmatched closing parentheses. The indices in the stack and the array
 * of unmatched closing parentheses are then merged and sorted. The function then constructs the
 * result string by skipping the characters at the indices in
 * the merged array.
 *
 * @param {string} s - The input string.
 * @returns {string} The input string with the minimum number of parentheses removed to make it valid.
 */
let minRemoveToMakeValid = function (s) {
    // Stack to keep track of the indices of the open parentheses
    let stack = [];
    // Array to keep track of the indices of the unmatched closing parentheses
    let remove = [];
    for (let i = 0; i < s.length; i++) {
        if ("(" === s[i]) {
            stack.push(i);
        } else if (")" === s[i]) {
            if (0 === stack.length) {
                remove.push(i);
            } else {
                stack.pop();
            }
        }
    }
    // Merge and sort the indices in the stack and the array of unmatched closing parentheses
    stack = stack.concat(remove);
    stack = stack.sort((a, b) => a - b);
    // Construct the result string by skipping the characters at the indices in the merged array
    let result = "";
    let j = 0;
    for (let i = 0; i < s.length; i++) {
        if (j < stack.length && stack[j] === i) {
            j++;
        } else {
            result += s[i];
        }
    }
    // Return the input string with the minimum number of parentheses removed to make it valid
    return result;
};
